### ISSUE TITLE
 AssertionError: `base_name` argument not specified, and could not automatically determine the name from the viewset, as it does not have a `.queryset` attribute.

### ISSUE THREAD

**Issue by s3xy4ngyc** [Axiologue/DjangoRestMultipleModels#12]
I keep on Getting this error when using MultipleModelAPIView:

>  AssertionError: `base_name` argument not specified, and could not automatically determine the name from the viewset, as it does not have a `.queryset` attribute.

Code:

```
#views.py
class CombinedSet(MultipleModelAPIView):
    flat = True
    queryList = (
        (Foo.objects.all(), FooSerializer),
        (Bar.objects.all(), BarSerializer),
    )

#serializers.py
class OtherSerializer(serializers.HyperlinkedModelSerializer):
    class Meta:
        model = Other


class FooSerializer(serializers.HyperlinkedModelSerializer):
    other = OtherSerializer(many=True, read_only=True)

    class Meta:
        model = Foo


class BarSerializer(serializers.HyperlinkedModelSerializer):
    other = OtherSerializer(many=True, read_only=True)

    class Meta:
        model = Bar

#models.py

class Other(models.Model):
    title = models.CharField(max_length=255)
    image = models.ImageField('image')

    def __unicode__(self):
        return self.title


class Generic(models.Model):
    title = models.CharField(max_length=255)
    other= models.ManyToManyField(Other, related_name='other', blank=True)

    def __unicode__(self):
        return self.title


class Foo(Generic):
    foo = models.IntegerField(default=0)


class Bar(Generic):
    bar = models.CharField(max_length=50)
```

**Comment by MattBroach** [comment]
Thanks for the report.  A few follow up questions, so I can try to reproduce the bug:
- Which versions of django and DRF are you using?
- Are you getting the error when accessing the view from the browsable API, from json, or both?

**Comment by MattBroach** [comment]
Some, I did some digging, and I think this might be a problem not with the `MultilpleModelAPIView`, but with your router config.  Here's the relevant lines from the [Django Rest Framework documentation](http://www.django-rest-framework.org/api-guide/routers/):


> Note: The base_name argument is used to specify the initial part of the view name pattern. In the example above, that's the user or account part.

> Typically you won't need to specify the base_name argument, but if you have a viewset where you've defined a custom get_queryset method, then the viewset may not have a .queryset attribute set. If you try to register that viewset you'll see an error like this:

> ``` 
> 'base_name' argument not specified, and could not automatically determine the name from the viewset, as it does not have a '.queryset' attribute.
> ```
> This means you'll need to explicitly set the base_name argument when registering the viewset, as it could not be automatically determined from the model name. 

Since the `MultipleModelAPIView` doesn't establish a queryset attribute (it doesn't use it), you'll need to manually specify you're `base_name` when adding it to a router.  Let me know if that catches the problem.

**Comment by srugano** [comment]
How would you advise adding it ? I've added it 
`router.register(r'inoutreports', InOutPatientsViewset, base_name='inoutreports')`

Now it gives me this : 


> Traceback:
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
>   119.                 resolver_match = resolver.resolve(request.path_info)
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in resolve
>   366.             for pattern in self.url_patterns:
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in url_patterns
>   402.         patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in urlconf_module
>   396.             self._urlconf_module = import_module(self.urlconf_name)
> File "/usr/lib/python2.7/importlib/__init__.py" in import_module
>   37.     __import__(name)
> File "/home/username/cmam/cmam/urls.py" in <module>
>   27.     url(r'^cmam/', include('cmam_app.urls')),
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/conf/urls/__init__.py" in include
>   33.         urlconf_module = import_module(urlconf_module)
> File "/usr/lib/python2.7/importlib/__init__.py" in import_module
>   37.     __import__(name)
> File "/home/stock/cmam/cmam_app/urls.py" in <module>
>   15.     url(r'^', include(router.urls)),
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in urls
>   79.             self._urls = patterns('', *self.get_urls())
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_urls
>   318.         default_urls = super(DefaultRouter, self).get_urls()
> File "/home/username/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_urls
>   259.                 view = viewset.as_view(mapping, **route.initkwargs)
> 
> Exception Type: TypeError at /cmam/inoutreports/
> Exception Value: as_view() takes exactly 1 argument (3 given)

**Comment by MattBroach** [comment]
So I think the `base_name` issue is only half of the problem.  The other issue is that regular generic DRF APIViews and Viewsets don't mix well.  Rather than using the built in `MultipleModelAPIView`, you need to use a view built with something like a `ModelViewSet` (see [this DRF issue](https://github.com/tomchristie/django-rest-framework/issues/933) which mentions the problem).  Luckily, DjangoRestMultipleModels also provides a mixin for you to use.  This code is untested and may have bugs, but try something like this:

```
from drf_multiple_model.mixins import MultipleModelMixin

from rest_framework.viewsets import ModelViewSet


class MyViewsetIView(ModelViewSet, GenericAPIView):
    queryList = (
        (Foo.objects.all(), FooSerializer),
        (Bar.objects.all(), BarSerializer),
    )

    def get(self, request, *args, **kwargs):
        return self.list(request, *args, **kwargs)
```

You'll still need to add the `base_name` kwarg to your router, but now you'd be using the proper view type.  Let me know how this works; if it's successful, I can build a base ViewSet into the library like I did with the MultipleModelAPIView.

**Comment by srugano** [comment]
Hi, first of all Thanks.
But unfortunately it doesn't work at all. Too bad.

**Comment by MattBroach** [comment]
Can you give me a little more information about what errors you got when it didn't work?  I'd like to get this working for viewsets as well as views.

**Comment by srugano** [comment]
Sorry I was kind of busy, but the problem is that in your tutorial, you didn't show how what `urls.py` should contain. I've tried the simple route : 
 
    # serilalizers.py
    class IncomingPatientSerializer(serializers.ModelSerializer):
        class Meta:
            model = IncomingPatientsReport
            fields = ('report','total_debut_semaine','transfert_interne','date_of_first_week_day')

    class OutgoingPatientSerializer(serializers.ModelSerializer):
        class Meta:
            model = OutgoingPatientsReport
            fields = ('report','gueri','deces','abandon', 'transfert_interne','date_of_first_week_day') 

    # views.py
    class InOutViewset(MultipleModelAPIView):

        queryList = [
            (IncomingPatientsReport.objects.all(), IncomingPatientSerializer),
            (OutgoingPatientsReport.objects.all(), OutgoingPatientSerializer),
        ]

    # urls.py 
    from rest_framework import routers
    router = routers.DefaultRouter()
    router.register(r'districts', DistrictCDSViewSet)   # these normal drf viewsets which works normal 
    router.register(r'inoutreport', InOutViewset)    #  without mentionning base_name

I get this error : 

```
Traceback:
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  119.                 resolver_match = resolver.resolve(request.path_info)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in resolve
  366.             for pattern in self.url_patterns:
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in url_patterns
  402.         patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in urlconf_module
  396.             self._urlconf_module = import_module(self.urlconf_name)
File "/usr/lib/python2.7/importlib/__init__.py" in import_module
  37.     __import__(name)
File "/home/stock/cmam/cmam/urls.py" in <module>
  27.     url(r'^cmam/', include('cmam_app.urls')),
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/conf/urls/__init__.py" in include
  33.         urlconf_module = import_module(urlconf_module)
File "/usr/lib/python2.7/importlib/__init__.py" in import_module
  37.     __import__(name)
File "/home/stock/cmam/cmam_app/urls.py" in <module>
  10. router.register(r'inoutreport', InOutViewset)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in register
  62.             base_name = self.get_default_base_name(viewset)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_default_base_name
  137.         assert queryset is not None, '`base_name` argument not specified, and could ' \

Exception Type: AssertionError at /cmam/inoutreports/
Exception Value: `base_name` argument not specified, and could not automatically determine the name from the viewset, as it does not have a `.queryset` attribute.
```
Ok.I got it that I need to provide the base_name argument. so I change in the `urls.py` file to : `router.register(r'inoutreport', InOutViewset, base_name='inoutreport')` and thus I get this error:

```
Traceback:
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  119.                 resolver_match = resolver.resolve(request.path_info)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in resolve
  366.             for pattern in self.url_patterns:
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in url_patterns
  402.         patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in urlconf_module
  396.             self._urlconf_module = import_module(self.urlconf_name)
File "/usr/lib/python2.7/importlib/__init__.py" in import_module
  37.     __import__(name)
File "/home/stock/cmam/cmam/urls.py" in <module>
  27.     url(r'^cmam/', include('cmam_app.urls')),
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/conf/urls/__init__.py" in include
  33.         urlconf_module = import_module(urlconf_module)
File "/usr/lib/python2.7/importlib/__init__.py" in import_module
  37.     __import__(name)
File "/home/stock/cmam/cmam_app/urls.py" in <module>
  14.     url(r'^', include(router.urls)),
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in urls
  81.             self._urls = self.get_urls()
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_urls
  355.         urls = super(DefaultRouter, self).get_urls()
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_urls
  261.                 view = viewset.as_view(mapping, **route.initkwargs)

Exception Type: TypeError at /cmam/inoutreports/
Exception Value: as_view() takes exactly 1 argument (3 given)
```
What's the problem then ? I think in your sample in the documentation you should write clearly how we're going to write the `route` in the `urls` file. I'm using the DRF 3.4.5, and I've tested the same error occurs on 3.0.5. Django 1.8.4.

**Comment by MattBroach** [comment]
That sample code is still using the `MultipleModelAPIView`.  Did you try it by instead creating a `ModelViewSet` with `MultipleModelMixin`, as I suggested above?

**Comment by srugano** [comment]
Can you show me how the viewset would then be implemented ?

**Comment by srugano** [comment]
Produces the same errors:

Either: 

```
Traceback:
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  119.                 resolver_match = resolver.resolve(request.path_info)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in resolve
  366.             for pattern in self.url_patterns:
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in url_patterns
  402.         patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/core/urlresolvers.py" in urlconf_module
  396.             self._urlconf_module = import_module(self.urlconf_name)
File "/usr/lib/python2.7/importlib/__init__.py" in import_module
  37.     __import__(name)
File "/home/stock/cmam/cmam/urls.py" in <module>
  27.     url(r'^cmam/', include('cmam_app.urls')),
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/django/conf/urls/__init__.py" in include
  33.         urlconf_module = import_module(urlconf_module)
File "/usr/lib/python2.7/importlib/__init__.py" in import_module
  37.     __import__(name)
File "/home/stock/cmam/cmam_app/urls.py" in <module>
  17.     url(r'^', include(router.urls)),
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in urls
  79.             self._urls = patterns('', *self.get_urls())
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_urls
  318.         default_urls = super(DefaultRouter, self).get_urls()
File "/home/stock/.virtualenvs/cmam/local/lib/python2.7/site-packages/rest_framework/routers.py" in get_urls
  259.                 view = viewset.as_view(mapping, **route.initkwargs)

Exception Type: TypeError at /cmam/inoutreport/
Exception Value: as_view() takes exactly 1 argument (3 given)
```

when it's:

```
class InOutViewset(MultipleModelMixin, GenericAPIView):
    queryList = (
        (IncomingPatientsReport.objects.all(), IncomingPatientSerializer),
        (OutgoingPatientsReport.objects.all(), OutgoingPatientSerializer),
        )
    def get(self, request, *args, **kwargs):
        return self.list(request, *args, **kwargs)
```
If I use `ModelViewset` in place of  `GenericAPIView`, I end up needing to providing the `get_queryset`, and `get_serilalizer_class` methods. But I thought your module is made aware of those two ?

I guess people end up doing something like this : http://stackoverflow.com/questions/24482817/django-rest-framework-combining-and-paginating-multiple-models

**Comment by mehwang** [comment]
I ran into this same issue, and had success with the following:
```python
from rest_frameworks.viewsets import GenericViewSet
from drf_multiple_model.mixins import MultipleModelMixin

class MultipleModelViewSet(MultipleModelMixin, GenericViewSet):
    queryList = [
        (CustomModelA.objects.all(), CustomModelASerializer, "custom_model_a"),
        (CustomModelB.objects.all(), CustomModelBSerializer, "custom_model_b"), 
    ]
```
Router register method still uses the base_name argument.

**Comment by MattBroach** [comment]
Thanks @mehwang! When I get a chance, I'll build a generic ViewSet into the library and update the docs.

**Comment by MattBroach** [comment]
The new, just released version 1.8 now contains a generic MultipleModelViewSet, exactly as described by @mehwang.  You still need to add the `base_name` argument when registering with a router (which is now covered in the docs).
